### DISCUSSION TITLE
Turborepo is not creating the node_modules inside the apps/*

### DISCUSSION THREAD

**Discussion by AyushNautiyalDeveloper** [vercel/turborepo#8039] [Help] [answered]
### Summary

My project look like this:

root (node_modules)
└── apps
   - └── generator(node_modules)

I am trying to run the npm i in my root I am expecting that the node_modules gets installed in my generator app. but it is only installing node_modules in my root. 

<img width="510" alt="image" src="https://github.com/vercel/turbo/assets/138557911/6f2eb2e0-6464-4d99-80f8-24d871fcc24e">

**Comment by anthonyshew** [comment, marked as answer]
This is the correct behavior. npm hoists dependencies to the root and resolves those dependencies using its resolution algorithm.

**Comment by AyushNautiyalDeveloper** [comment]
@anthonyshew  I mean I have used turborepo for other projects also generally I have seen it installs node_modules for every app it have. Is it configured some how. If yes can you clearify. 
<img width="515" alt="image" src="https://github.com/vercel/turbo/assets/138557911/3a380158-e32c-4a8a-afef-186a6885534d">

**Reply by anthonyshew** [reply]
Are the dependencies resolving correctly? Can you build and develop your application?

If so, then everything is working as expected. Different package managers handle dependencies in different ways. Additionally, depending on how/where the dependencies you are using are installed, you can end up with `node_modules` in different places.

**Reply by ayushnau** [reply]
1. ok but I need node_modules in my local can i state that explicitly
for the dependencies. also no due to some issue in dependencies the project is not able to run tests. which is the main issue i am having .
2. Also Is it not a good idea to use package.lock.json inside the apps/generator (in my case).
3. Also can you repoen this. since it is still active.

 @anthonyshew

**Reply by anthonyshew** [reply]
You may be able to tune the behavior of where things get installed for your use case using [an `.npmrc`](https://docs.npmjs.com/cli/v10/configuring-npm/npmrc). At a glance, [`install-strategy`](https://docs.npmjs.com/cli/v10/using-npm/config#install-strategy) looks like it may help with what 

You don't want to have a `package-lock.json` in your packages. There should only be one in the root of your repository.

I'd prefer to keep this closed because the top-level question is answered to the level that will benefit the community. Additionally,
-  This isn't a question about Turborepo. Dependency resolution is handled by package managers. But I know it can be frustrating when it goes wrong so trying to help out, anyway!
- The issue appears to be either a misconfiguration in the repository or highly-specific to your use case.

**Reply by mehulkar** [reply]
@ayushnau I'm happy can re-open this if you can provide a reproduction. I agree with Anthony that this doesn't look like it has anything to do with Turbo. You may be able to find more help in the [turbo discord as a support channel](https://turbo.build/discord).

**Reply by mehulkar** [reply]
> also no due to some issue in dependencies the project is not able to run tests. which is the main issue i am having .

Please provide some kind of evidence that this is related to turbo or a reproduction.

**Reply by AyushNautiyalDeveloper** [reply]
will do that  @mehulkar

**Reply by AyushNautiyalDeveloper** [reply]
@mehulkar I have uploaded my project here please check this. https://github.com/ayushnau/generator

run npm i and see for yourself the node_modules not getting installed in generator repo. 
due to which test are getting failed.

**Reply by mehulkar** [reply]
@ayushnau thanks for the reproduction. I cloned it down and am at the same conclusion: that your issue is not related to Turborepo. Some notes again:

- `npm` can install your dependencies wherever it likes. turbo cannot control this, and in my experience it's not necessarily a bug that there after `npm i` `apps/generator/node_modules` does not exist. I am 99% sure this is intentional.

- I attempted to replace your reproduction with `pnpm@8.15.0` instead, which _does_ create `apps/generator/node_modules` and your tests still fail the same way. I also tried with Node 18 and Node 20, since the error seemed related to some core built ins

    <details>
    <summary>Stack trace</summary>

    ```logs
    PASS  test/renderer.test.js
    FAIL  test/utils.test.js
      ● Test suite failed to run

        ENOENT: no such file or directory, open 'node:path'

        > 1 | const npmConfig = jest.genMockFromModule('@npmcli/config');
            |                                                           ^
          2 |
          3 | module.exports = npmConfig;
          4 |

          at Runtime.readFile (../../node_modules/jest-runtime/build/index.js:1842:21)
          at Object.<anonymous> (../../node_modules/.pnpm/@npmcli+config@8.3.1/node_modules/@npmcli/config/lib/index.js:7:36)
          at Object.<anonymous> (test/__mocks__/@npmcli/config.js:1:59)
          at Object.<anonymous> (lib/generator.js:10:16)
          at Object.<anonymous> (test/utils.test.js:3:19)

    FAIL  test/generator.test.js
      ● Test suite failed to run

        ENOENT: no such file or directory, open 'node:path'

        > 1 | const npmConfig = jest.genMockFromModule('@npmcli/config');
            |                                                           ^
          2 |
          3 | module.exports = npmConfig;
          4 |

          at Runtime.readFile (../../node_modules/jest-runtime/build/index.js:1842:21)
          at Object.<anonymous> (../../node_modules/.pnpm/@npmcli+config@8.3.1/node_modules/@npmcli/config/lib/index.js:7:36)
          at Object.<anonymous> (test/__mocks__/@npmcli/config.js:1:59)
          at Object.<anonymous> (lib/generator.js:10:16)
          at Object.<anonymous> (test/generator.test.js:4:19)
    ```
    </details>


Even if this _is_ some kind of bug with something outside your code, I am reasonably confident at this point that Turborepo's existence in your repo is not causing this. You can try removing Turborepo entirely from your repo, or run your tests directly with `jest` to confirm this for yourself.

**Reply by maxondev-01001** [reply]
i can run scripts using npm

i have nm version 10 and 20 node version

**Reply by maxondev-01001** [reply]
how i an fix this error?
